client/client-lib/src/api.rs

```diff
@@ -198,3 +275,234 @@ impl<C: ClientT> WsFederationApi<C> {
         ))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use anyhow::anyhow;
+    use lazy_static::lazy_static;
+    use std::{
+        collections::HashSet,
+        sync::{
+            atomic::{AtomicBool, AtomicUsize, Ordering},
+            Mutex,
+        },
+    };
+
+    use super::*;
+
+    type Result<T = ()> = std::result::Result<T, JsonRpcError>;
+
+    #[async_trait]
+    trait SimpleClient: Sized {
+        async fn connect() -> Result<Self>;
+        fn is_connected(&self) -> bool {
+            true
+        }
+        // reply with json
+        async fn request(&self, method: &str) -> Result<String>;
+    }
+
+    struct Client<C: SimpleClient>(C);
+
+    #[async_trait]
+    impl<C: SimpleClient + Send + Sync> JsonRpcClient for Client<C> {
+        fn is_connected(&self) -> bool {
+            self.0.is_connected()
+        }
+
+        async fn connect(_url: &str) -> std::result::Result<Self, JsonRpcError> {
+            Ok(Self(C::connect().await?))
+        }
+    }
+
+    #[async_trait]
+    impl<C: SimpleClient + Send + Sync> ClientT for Client<C> {
+        async fn request<'a, R>(
+            &self,
+            method: &'a str,
+            _params: Option<jsonrpsee_types::ParamsSer<'a>>,
+        ) -> std::result::Result<R, JsonRpcError>
+        where
+            R: jsonrpsee_core::DeserializeOwned,
+        {
+            let json = self.0.request(method).await?;
+            Ok(serde_json::from_str(&json).unwrap())
+        }
+
+        async fn notification<'a>(
+            &self,
+            _method: &'a str,
+            _params: Option<jsonrpsee_types::ParamsSer<'a>>,
+        ) -> std::result::Result<(), JsonRpcError> {
+            unimplemented!()
+        }
+
+        async fn batch_request<'a, R>(
+            &self,
+            _batch: Vec<(&'a str, Option<jsonrpsee_types::ParamsSer<'a>>)>,
+        ) -> std::result::Result<Vec<R>, JsonRpcError>
+        where
+            R: jsonrpsee_core::DeserializeOwned + Default + Clone,
+        {
+            unimplemented!()
+        }
+    }
+
+    fn federation_member<C: SimpleClient + Send + Sync>() -> FederationMember<Client<C>> {
+        FederationMember {
+            url: String::new(),
+            peer_id: PeerId::from(0),
+            client: RwLock::new(None),
+        }
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn test_connect() {
+        static CONNECTION_COUNT: AtomicUsize = AtomicUsize::new(0);
+        static CONNECTED: AtomicBool = AtomicBool::new(true);
```

Comment:
Why make these static? Could be an `Arc` stored in `Client` imo.

Reply:
because Client::connect doesn't have `self`

Reply:
Right, I was still in the trait object mindset where you always have self, but that would require two traits here to be object safe, so let's keep it like that. Sorry for getting confused.